IPs/RSKIP387.md

```diff
@@ -0,0 +1,303 @@
+
+The UTXOs will be owned in RSK by the same private key that controls the first input of the transaction or owned by a different RSK address, if there is an `OP_RETURN` output containing it, as specified by RSKIP-170.
+
+The method call must transfer the full payment of the annual parking cost of all the registered ordinals, as if `payParkingExtensionFee()` would have been called for each ordinal. If the amount paid is less than the amount required, the transaction will fail and the amount paid will be reimbursed to the calling address. If the amount is greated than the required, the excess will be reimbursed.
+
+
+## Peg-In using Destination-Mapped Federation Addresses
+
+It’s also possible to create a Bitcoin address that automatically assigns the ownership of the ordinal in Rootstock to a specific RSK address or contract. This is called destination-mapped federation addresses (DMFA). This can be used transfer ordinals when the source wallet doen't allow adding an `OP_RETURN` output, nor controlling the private key associated with the first input.
+
+This requires creating a special unique single-use reception address for the ordinal. The transaction is registered with the following method:
+
+
+`function registerBtcTransactionWithOrdinalsDMFA(
+   bytes btcTxSerialized,
+   uint32 height,
+   bytes pmtSerialized,
+   address rskOwner);`
+
+The field rskOwner specifies the owner of the ordinal.
+
+## Changing the Ownership of an Ordinal
+
+To query who owns an ordinal, the following method is added:
+
+`function getOrdinalOwner(uint256 tId, int index) returns (address)`
+
+The method will return 0x00... (the zero 20-byte address) if the ordinal does not exist.
+
+To change ownership of an ordinal, the following method must be called by the owner:
+
+`function transferOrdinal(uint256 tId, int index, address newOwner)`
+
+The method does nothing if the ordinal does not exist.
+
+## Ordinal Peg-outs
+
+To send an ordinal, the following method must be called:
+
+`function releaseOrdinal(uint256 txId, uint256 index, bytes btcScriptPubkey)`
+
+This method will queue the release of an ordinal, sending it to the Bitcoin address associated with the btcScriptPubkey specified. The caller must be the owner of the ordinal. Every 1 hour the ordinals that are queued to be released will be processed. The bridge may attempt to create a single transaction containing all ordinal UTXOs, and all ordinal outputs, plus a single additional input used for fee payments, and a single additional output used for change.
+
+Ordinals can be sent to any Bitcoin address by directly specifying the btcScriptPubkey, and the called must make sure that the btcScriptPubkey given is the correct expansion of a standard, Bech32 or any other future Bitcoin address. The Bridge will not perform any validation on the btcScriptPubkey. If it is invalid, the ordinal will be lost. 
```

Comment:
I see a problem with this and peg-out batching: what if the btcScriptPubkey makes the transaction non-standard ? We would need to perform some minimal validation on the scriptPub to avoid it.

Reply:
@josedahlquist Do you agree?